pudb/debugger.py

```diff
@@ -216,34 +216,36 @@ def set_trace(self, frame=None, as_breakpoint=None, paused=True):
                 as_breakpoint = True
 
         if frame is None:
-            frame = thisframe = sys._getframe().f_back
-        else:
-            thisframe = frame
+            frame = sys._getframe(1)
+
         # See pudb issue #52. If this works well enough we should upstream to
         # stdlib bdb.py.
         #self.reset()
 
-        while frame:
-            frame.f_trace = self.trace_dispatch
-            self.botframe = frame
-            frame = frame.f_back
-
-        thisframe_info = (
-                self.canonic(thisframe.f_code.co_filename), thisframe.f_lineno)
-        if thisframe_info not in self.set_traces or self.set_traces[thisframe_info]:
-            if as_breakpoint:
-                self.set_traces[thisframe_info] = True
-                if self.ui.source_code_provider is not None:
-                    self.ui.set_source_code_provider(
-                            self.ui.source_code_provider, force_update=True)
-
-            if paused:
-                self.set_step()
-            else:
-                self.set_continue()
-            sys.settrace(self.trace_dispatch)
-        else:
-            return
+        self.botframe = frame
+        self.botframe.f_trace = self.trace_dispatch
+        while self.botframe.f_back:
+            self.botframe = self.botframe.f_back
+            self.botframe.f_trace = self.trace_dispatch
+
+        stack, _ = self.get_stack(frame, None)
+        if stack:
```

Comment:
OK, the check is fine. I agree it shouldn't happen, unless you do something really weird (like inject a set_trace in the importlib._bootstrap code). Might make the code easier to read if you inverted the check to avoid indenting the whole block:

```py
if not stack:
    return
...
```